harness/determined/launch/autodeepspeed.py

```diff
@@ -0,0 +1,256 @@
+            "3",
+            f"/tmp/pid_server-{info.allocation_id}",
+            str(len(info.slot_ids)),
+            "--",
+        ]
+
+        logging.debug(
+            f"Non-chief [{info.container_rank}] training process launch "
+            f"command: {run_sshd_command}."
+        )
+        return subprocess.Popen(pid_server_cmd + run_sshd_command).wait()
+
+    # We always need to set this variable to initialize the context correctly, even in the single
+    # slot case.
+    os.environ["USE_DEEPSPEED"] = "1"
+
+    if len(info.container_addrs) > 1:
+        # Create the environment file that will be passed by deepspeed to individual ranks.
+        create_deepspeed_env_file()
+        # Set custom PDSH args:
+        # * bypass strict host checking
+        # * -p our custom port
+        # * other args are default ssh args for pdsh
+        os.environ["PDSH_SSH_ARGS"] = (
+            "-o PasswordAuthentication=no -o StrictHostKeyChecking=no "
+            f"-p {constants.DTRAIN_SSH_PORT} -2 -a -x %h"
+        )
+
+        # Chief worker also needs to run sshd when using pdsh.
+        subprocess.Popen(run_sshd_command)
+        # Once the launch process wrapped in the pid_server finishes, the container will still have
+        # a daemon sshd process running.  Hence, we'll mark the chief  containers as a daemon
+        # as well so that the master can cleanup after the pid_server command below exists.
+        api.post(
+            info.master_url,
+            path=f"/api/v1/allocations/{info.allocation_id}/containers/{container_id}/daemon",
+            cert=cert,
+        )
+
+        # Chief machine waits for every worker's sshd to be available.  All machines should be
+        # close to in-step by now because all machines just finished synchronizing rendezvous info.
+        deadline = time.time() + 20
+        for peer in info.container_addrs:
+            while True:
+                with socket.socket() as sock:
+                    sock.settimeout(1)
+                    try:
+                        # Connect to a socket to ensure sshd is listening.
+                        sock.connect((peer, constants.DTRAIN_SSH_PORT))
+                        # The ssh protocol requires the server to serve an initial greeting.
+                        # Receive part of that greeting to know that sshd is accepting/responding.
+                        data = sock.recv(1)
+                        if not data:
+                            raise ValueError("no sshd greeting")
+                        # This peer is ready.
+                        break
+                    except Exception:
+                        if time.time() > deadline:
+                            raise ValueError(
+                                f"Chief machine was unable to connect to sshd on peer machine at "
+                                f"{peer}:{constants.DTRAIN_SSH_PORT}"
+                            )
+                        time.sleep(0.1)
```

Comment:
can you make this a helper function in `det.util` maybe, and call it from both here and autohorovod?

I think @azhou-determined is who I need to tag here, but I'm not sure.